test/core/bad_connection/close_fd_test.cc

```diff
@@ -0,0 +1,731 @@
+  GPR_ASSERT(ctx->client);
+  grpc_chttp2_transport_start_reading(transport, nullptr, nullptr);
+}
+
+static void init_server(test_ctx* ctx) {
+  grpc_core::ExecCtx exec_ctx;
+  grpc_transport* transport;
+  GPR_ASSERT(!ctx->server);
+  ctx->server = grpc_server_create(nullptr, nullptr);
+  grpc_server_register_completion_queue(ctx->server, ctx->cq, nullptr);
+  grpc_server_start(ctx->server);
+  transport = grpc_create_chttp2_transport(nullptr, ctx->ep->server, false);
+  server_setup_transport(ctx, transport);
+  grpc_chttp2_transport_start_reading(transport, nullptr, nullptr);
+}
+
+static void test_init(test_ctx* ctx) {
+  grpc_endpoint_pair* sfd =
+      static_cast<grpc_endpoint_pair*>(gpr_malloc(sizeof(grpc_endpoint_pair)));
+  memset(ctx, 0, sizeof(*ctx));
+  ctx->ep = sfd;
+  ctx->cq = grpc_completion_queue_create_for_next(nullptr);
+  ctx->shutdown_cq = grpc_completion_queue_create_for_pluck(nullptr);
+  ctx->bound_cq = grpc_completion_queue_create_for_next(nullptr);
+  ctx->client_cq = grpc_completion_queue_create_for_next(nullptr);
+
+  /* Create endpoints */
+  *sfd = grpc_iomgr_create_endpoint_pair("fixture", nullptr);
+  /* Create client, server and setup transport over endpoint pair */
+  init_server(ctx);
+  init_client(ctx);
+}
+
+static void drain_cq(grpc_completion_queue* cq) {
+  grpc_event ev;
+  do {
+    ev = grpc_completion_queue_next(cq, grpc_timeout_seconds_to_deadline(1),
+                                    nullptr);
+  } while (ev.type != GRPC_QUEUE_SHUTDOWN);
+}
+
+static void drain_and_destroy_cq(grpc_completion_queue* cq) {
+  grpc_completion_queue_shutdown(cq);
+  drain_cq(cq);
+  grpc_completion_queue_destroy(cq);
+}
+
+static void shutdown_server(test_ctx* ctx) {
+  if (!ctx->server) return;
+  grpc_server_shutdown_and_notify(ctx->server, ctx->shutdown_cq, tag(1000));
+  GPR_ASSERT(grpc_completion_queue_pluck(ctx->shutdown_cq, tag(1000),
+                                         grpc_timeout_seconds_to_deadline(1),
+                                         nullptr)
+                 .type == GRPC_OP_COMPLETE);
+  grpc_server_destroy(ctx->server);
+  ctx->server = nullptr;
+}
+
+static void shutdown_client(test_ctx* ctx) {
+  if (!ctx->client) return;
+  grpc_channel_destroy(ctx->client);
+  ctx->client = nullptr;
+}
+
+static void end_test(test_ctx* ctx) {
+  shutdown_server(ctx);
+  shutdown_client(ctx);
+
+  drain_and_destroy_cq(ctx->cq);
+  drain_and_destroy_cq(ctx->client_cq);
+  drain_and_destroy_cq(ctx->bound_cq);
+  grpc_completion_queue_destroy(ctx->shutdown_cq);
+  gpr_free(ctx->ep);
+}
+
+typedef enum fd_type { CLIENT_FD, SERVER_FD } fd_type;
+
+static const char* fd_type_str(fd_type f) {
+  if (f == CLIENT_FD) {
+    return "client";
+  } else if (f == SERVER_FD) {
+    return "server";
+  } else {
+    gpr_log(GPR_ERROR, "Unexpected fd_type %d", f);
+    abort();
+  }
+}
+
+static void _test_close_before_server_recv(fd_type f) {
+  grpc_core::ExecCtx exec_ctx;
+  grpc_call* c;
```

Comment:
Try avoid single letter identifier like this. Not quite readable in the code, and maybe not comply with style guide

Reply:
Fixed